rootio/wbuffer.go

```diff
@@ -4,6 +4,154 @@
 
 package rootio
 
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+type wbuff struct {
+	p []byte // buffer of data to write on
+	c int    // current position in buffer of data
+}
+
+func (w *wbuff) Write(p []byte) (int, error) {
+	if w.c >= len(w.p) {
+		return 0, io.EOF
+	}
+	n := copy(w.p[w.c:], p)
+	w.c += n
+	return n, nil
+}
+
+func (w *wbuff) WriteByte(p byte) error {
+	if w.c >= len(w.p) {
+		return io.EOF
+	}
+	w.p[w.c] = p
+	w.c++
+	return nil
+}
+
 // WBuffer is a write-only ROOT buffer for streaming.
 type WBuffer struct {
+	w      *wbuff
+	err    error
+	offset uint32
+	refs   map[int64]interface{}
+	sictx  StreamerInfoContext
+}
+
+func NewWBuffer(data []byte, refs map[int64]interface{}, offset uint32, ctx StreamerInfoContext) *WBuffer {
+	if refs == nil {
+		refs = make(map[int64]interface{})
+	}
+
+	return &WBuffer{
+		w:      &wbuff{p: data, c: 0},
+		refs:   refs,
+		offset: offset,
+		sictx:  ctx,
+	}
+}
+
+func (w *WBuffer) WriteI8(i int8) {
+	if w.err != nil {
+		return
+	}
+	v := byte(i)
+	w.w.WriteByte(v)
+}
+
+func (w *WBuffer) WriteI16(i int16) {
+	if w.err != nil {
+		return
+	}
+	binary.BigEndian.PutUint16(w.w.p, uint16(i))
```

Comment:
here, you don't check whether there is enough room inside the `w.w.p` buffer for `i`.
also, you write it at the beginning of that `p` slice of bytes.
you should instead write it at the current cursor location (_ie_ `w.w.c`)
and, after the write, you didn't update the current cursor location (_ie_, it should have advanced by 2 bytes.)

_e.g._:
```go
beg := w.w.c
end := w.w.c + 2
binary.BigEndian.PutUint16(w.w.p[beg:end], uint16(i))
w.w.c += 2
```